### PR TITLE
shader_cache: Fix use-after-free and orphan invalidation cache entries

### PR DESCRIPTION
**Pull request by ReinUsesLisp**
This fixes some cases where entries could have been removed multiple
times reading freed memory. To address this issue this commit removes
duplicates from entries marked for removal and sorts out the removal
process to fix another use-after-free situation.

Another issue fixed in this commit is orphan invalidation cache entries.
Previously only the entries that were invalidated in the current
operations had its entries removed. This led to more use-after-free
situations when these entries were actually invalidated but referenced
an object that didn't exist.

- Fixes vertex explosions on Xenoblade Chronicles 2
- It may fix shader issues on Smash Ultimate

### CODE REVIEW

File: src/video_core/shader_cache.h [hunk 1]
```diff
@@ -20,6 +20,7 @@ namespace VideoCommon {
 template <class T>
 class ShaderCache {
     static constexpr u64 PAGE_BITS = 14;
+    static constexpr u64 PAGE_SIZE = u64(1) << PAGE_BITS;
```
Comment (ogniK5377) [inline src/video_core/shader_cache.h:23]:
```suggestion
    static constexpr u64 PAGE_SIZE = 1UUL << PAGE_BITS;
```

Comment (Morph1984) [reply src/video_core/shader_cache.h:23]: UUL or ULL

Comment (VolcaEM) [reply src/video_core/shader_cache.h:23]: unsigned long long so I guess ULL

Comment (FernandoS27) [reply src/video_core/shader_cache.h:23]: Well it turns out Unsigned long long does not mean the same in linux and windows, so his approach is safer.

**Comment by ogniK5377** [comment]
Untagged due to softlocks being caused in pokemon

**Comment by ReinUsesLisp** [comment]
Pushed a fix for it, there was a bug in `InvalidatePageEntries`. `index` was being increased when it shouldn't.
I played 15 battles on Pokémon Sword without experiencing a softlock.

Here's the diff:
```diff
diff --git a/src/video_core/shader_cache.h b/src/video_core/shader_cache.h
index 4a8fe7c2a..b7608fc7b 100644
--- a/src/video_core/shader_cache.h
+++ b/src/video_core/shader_cache.h
@@ -155,7 +155,8 @@ private:
     /// @param addr_end        Non-inclusive end address of the invalidation
     /// @pre invalidation_mutex is locked
     void InvalidatePageEntries(std::vector<Entry*>& entries, VAddr addr, VAddr addr_end) {
-        for (std::size_t index = 0; index < entries.size(); ++index) {
+        std::size_t index = 0;
+        while (index < entries.size()) {
             Entry* const entry = entries[index];
             if (!entry->Overlaps(addr, addr_end)) {
                 ++index;
```

**Pull request merged**